File: src/main/java/org/alpacaindustries/iremiaminigame/command/MinigameCommand.java
package org.alpacaindustries.iremiaminigame.command;

import org.alpacaindustries.iremiaminigame.IremiaMinigameCorePlugin;
import org.alpacaindustries.iremiaminigame.minigame.Minigame;
import org.alpacaindustries.iremiaminigame.minigame.MinigameManager;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.Location;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Command handler for minigame commands using the traditional Bukkit API
 * This approach avoids type issues with Paper's Brigadier implementation
 */
public class MinigameCommand implements CommandExecutor, TabCompleter {
  private final IremiaMinigameCorePlugin plugin;
  private final MinigameManager minigameManager;
  private static final Component PREFIX = Component.text("[", NamedTextColor.DARK_GRAY)
      .append(Component.text("Minigame", NamedTextColor.GOLD, TextDecoration.BOLD))
      .append(Component.text("] ", NamedTextColor.DARK_GRAY));

  public MinigameCommand(IremiaMinigameCorePlugin plugin) {
    this.plugin = plugin;
    this.minigameManager = plugin.getMinigameManager();

    // Register command the traditional way
    plugin.getCommand("minigame").setExecutor(this);
    plugin.getCommand("minigame").setTabCompleter(this);
  }

  @Override
  public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
    if (args.length == 0) {
      showHelp(sender);
      return true;
    }

    String subCommand = args[0].toLowerCase();

    switch (subCommand) {
      case "join":
        return handleJoinCommand(sender, args);
      case "leave":
        return handleLeaveCommand(sender, args);
      case "create":
        return handleCreateCommand(sender, args);
      case "list":
        return handleListCommand(sender, args);
      case "start":
        return handleStartCommand(sender, args);
      case "end":
        return handleEndCommand(sender, args);
      case "setmin":
        return handleSetMinCommand(sender, args);
      case "setmax":
        return handleSetMaxCommand(sender, args);
      case "setspawn":
        return handleSetSpawnCommand(sender, args);
      case "help":
      default:
        showHelp(sender);
        return true;
    }
  }

  @Override
  public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
    List<String> completions = new ArrayList<>();

    if (args.length == 1) {
      // Main subcommands
      List<String> subCommands = new ArrayList<>();
      subCommands.add("join");
      subCommands.add("leave");
      subCommands.add("list");
      subCommands.add("help");

      // Add admin commands if player has permission
      if (sender.hasPermission("iremiaminigame.command.create")) {
        subCommands.add("create");
      }
      if (sender.hasPermission("iremiaminigame.command.start")) {
        subCommands.add("start");
      }
      if (sender.hasPermission("iremiaminigame.command.end")) {
        subCommands.add("end");
      }
      if (sender.hasPermission("iremiaminigame.command.setmin")) {
        subCommands.add("setmin");
      }
      if (sender.hasPermission("iremiaminigame.command.setmax")) {
        subCommands.add("setmax");
      }
      if (sender.hasPermission("iremiaminigame.command.setspawn")) {
        subCommands.add("setspawn");
      }

      return filterCompletions(subCommands, args[0]);
    }

    if (args.length == 2) {
      switch (args[0].toLowerCase()) {
        case "join":
        case "start":
        case "end":
        case "setmin":
        case "setmax":
        case "setspawn":
          // Show active games
          return filterCompletions(new ArrayList<>(minigameManager.getActiveGames().keySet()), args[1]);
        case "create":
          // Show available game types
          return filterCompletions(new ArrayList<>(minigameManager.getMinigameTypes()), args[1]);
      }
    }

    if (args.length == 3) {
      if (args[0].toLowerCase().equals("create")) {
        // For create command, suggest some default IDs
        completions.add("game1");
        completions.add("game2");
        completions.add("custom");
        completions.add(args[1] + "_1");
        return filterCompletions(completions, args[2]);
      }
    }

    return completions;
  }

  /**
   * Filter completions by prefix
   */
  private List<String> filterCompletions(List<String> possibilities, String prefix) {
    if (prefix.isEmpty()) {
      return possibilities;
    }

    prefix = prefix.toLowerCase();
    List<String> filtered = new ArrayList<>();
    for (String option : possibilities) {
      if (option.toLowerCase().startsWith(prefix)) {
        filtered.add(option);
      }
    }
    return filtered;
  }

  // Command handlers

  private boolean handleJoinCommand(CommandSender sender, String[] args) {
    if (!(sender instanceof Player player)) {
      sender.sendMessage(PREFIX.append(Component.text("Only players can join games.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 2) {
      player.sendMessage(PREFIX.append(Component.text("Usage: /minigame join <gameId>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    if (minigameManager.addPlayerToGame(player, gameId)) {
      player.sendMessage(PREFIX.append(Component.text("You joined the game: ", NamedTextColor.GREEN))
          .append(Component.text(gameId, NamedTextColor.GOLD)));
    } else {
      player.sendMessage(PREFIX.append(Component.text("Failed to join the game.", NamedTextColor.RED)));
    }

    return true;
  }

  private boolean handleLeaveCommand(CommandSender sender, String[] args) {
    if (!(sender instanceof Player player)) {
      sender.sendMessage(PREFIX.append(Component.text("Only players can leave games.", NamedTextColor.RED)));
      return true;
    }

    if (minigameManager.removePlayerFromGame(player)) {
      player.sendMessage(PREFIX.append(Component.text("You left the game.", NamedTextColor.GREEN)));
    } else {
      player.sendMessage(PREFIX.append(Component.text("You are not in a game.", NamedTextColor.RED)));
    }

    return true;
  }

  private boolean handleCreateCommand(CommandSender sender, String[] args) {
    if (!sender.hasPermission("iremiaminigame.command.create")) {
      sender.sendMessage(
          PREFIX.append(
              Component.text("You don't have permission to create minigames.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 3) {
      sender.sendMessage(
          PREFIX.append(Component.text("Usage: /minigame create <type> <id>", NamedTextColor.RED)));
      return true;
    }

    String gameType = args[1];
    String gameId = args[2];

    Minigame minigame = minigameManager.createMinigame(gameType, gameId);
    if (minigame != null) {
      sender.sendMessage(PREFIX.append(Component.text("Created game: ", NamedTextColor.GREEN))
          .append(Component.text(minigame.getDisplayName() + " (" + gameId + ")", NamedTextColor.GOLD)));
    } else {
      sender.sendMessage(PREFIX.append(Component.text("Failed to create game.", NamedTextColor.RED)));
    }

    return true;
  }

  private boolean handleListCommand(CommandSender sender, String[] args) {
    Map<String, Minigame> games = minigameManager.getActiveGames();

    if (games.isEmpty()) {
      sender.sendMessage(PREFIX.append(Component.text("There are no active games.", NamedTextColor.YELLOW)));
      return true;
    }

    sender.sendMessage(PREFIX.append(Component.text("Active Games:", NamedTextColor.YELLOW)));
    for (Map.Entry<String, Minigame> entry : games.entrySet()) {
      Minigame game = entry.getValue();
      sender.sendMessage(Component.text("- ", NamedTextColor.GRAY)
          .append(Component.text(entry.getKey(), NamedTextColor.GOLD))
          .append(Component.text(" (" + game.getDisplayName() + "): ", NamedTextColor.YELLOW))
          .append(Component.text(game.getPlayerCount() + " players, ", NamedTextColor.WHITE))
          .append(Component.text("Status: " + game.getState(), NamedTextColor.AQUA)));
    }

    return true;
  }

  private boolean handleStartCommand(CommandSender sender, String[] args) {
    if (!sender.hasPermission("iremiaminigame.command.start")) {
      sender.sendMessage(
          PREFIX.append(Component.text("You don't have permission to start minigames.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 2) {
      sender.sendMessage(PREFIX.append(Component.text("Usage: /minigame start <gameId>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    Minigame game = minigameManager.getActiveGames().get(gameId);

    if (game == null) {
      sender.sendMessage(PREFIX.append(Component.text("Game not found: " + gameId, NamedTextColor.RED)));
      return true;
    }

    game.startCountdown();
    sender.sendMessage(PREFIX.append(Component.text("Starting countdown for game: ", NamedTextColor.GREEN))
        .append(Component.text(gameId, NamedTextColor.GOLD)));

    return true;
  }

  private boolean handleEndCommand(CommandSender sender, String[] args) {
    if (!sender.hasPermission("iremiaminigame.command.end")) {
      sender.sendMessage(
          PREFIX.append(Component.text("You don't have permission to end minigames.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 2) {
      sender.sendMessage(PREFIX.append(Component.text("Usage: /minigame end <gameId>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    Minigame game = minigameManager.getActiveGames().get(gameId);

    if (game == null) {
      sender.sendMessage(PREFIX.append(Component.text("Game not found: " + gameId, NamedTextColor.RED)));
      return true;
    }

    game.end();
    sender.sendMessage(PREFIX.append(Component.text("Ended game: ", NamedTextColor.GREEN))
        .append(Component.text(gameId, NamedTextColor.GOLD)));

    return true;
  }

  private boolean handleSetMinCommand(CommandSender sender, String[] args) {
    if (!sender.hasPermission("iremiaminigame.command.setmin")) {
      sender.sendMessage(
          PREFIX.append(
              Component.text("You don't have permission to set minimum players.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 3) {
      sender.sendMessage(
          PREFIX.append(Component.text("Usage: /minigame setmin <gameId> <minPlayers>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    Minigame game = minigameManager.getActiveGames().get(gameId);

    if (game == null) {
      sender.sendMessage(PREFIX.append(Component.text("Game not found: " + gameId, NamedTextColor.RED)));
      return true;
    }

    try {
      int minPlayers = Integer.parseInt(args[2]);
      if (minPlayers < 1) {
        sender.sendMessage(
            PREFIX.append(Component.text("Minimum players must be at least 1.", NamedTextColor.RED)));
        return true;
      }

      game.setMinPlayers(minPlayers);
      sender.sendMessage(
          PREFIX.append(Component.text("Set minimum players to " + minPlayers + " for game: ",
              NamedTextColor.GREEN))
              .append(Component.text(gameId, NamedTextColor.GOLD)));
    } catch (NumberFormatException e) {
      sender.sendMessage(PREFIX.append(Component.text("Invalid number: " + args[2], NamedTextColor.RED)));
    }

    return true;
  }

  private boolean handleSetMaxCommand(CommandSender sender, String[] args) {
    if (!sender.hasPermission("iremiaminigame.command.setmax")) {
      sender.sendMessage(
          PREFIX.append(
              Component.text("You don't have permission to set maximum players.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 3) {
      sender.sendMessage(
          PREFIX.append(Component.text("Usage: /minigame setmax <gameId> <maxPlayers>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    Minigame game = minigameManager.getActiveGames().get(gameId);

    if (game == null) {
      sender.sendMessage(PREFIX.append(Component.text("Game not found: " + gameId, NamedTextColor.RED)));
      return true;
    }

    try {
      int maxPlayers = Integer.parseInt(args[2]);
      if (maxPlayers < 1) {
        sender.sendMessage(
            PREFIX.append(Component.text("Maximum players must be at least 1.", NamedTextColor.RED)));
        return true;
      }

      game.setMaxPlayers(maxPlayers);
      sender.sendMessage(
          PREFIX.append(Component.text("Set maximum players to " + maxPlayers + " for game: ",
              NamedTextColor.GREEN))
              .append(Component.text(gameId, NamedTextColor.GOLD)));
    } catch (NumberFormatException e) {
      sender.sendMessage(PREFIX.append(Component.text("Invalid number: " + args[2], NamedTextColor.RED)));
    }

    return true;
  }

  private boolean handleSetSpawnCommand(CommandSender sender, String[] args) {
    if (!(sender instanceof Player player)) {
      sender.sendMessage(
          PREFIX.append(Component.text("Only players can set spawn locations.", NamedTextColor.RED)));
      return true;
    }

    if (!sender.hasPermission("iremiaminigame.command.setspawn")) {
      sender.sendMessage(
          PREFIX.append(
              Component.text("You don't have permission to set spawn points.", NamedTextColor.RED)));
      return true;
    }

    if (args.length < 2) {
      sender.sendMessage(PREFIX.append(Component.text("Usage: /minigame setspawn <gameId>", NamedTextColor.RED)));
      return true;
    }

    String gameId = args[1];
    Minigame game = minigameManager.getActiveGames().get(gameId);

    if (game == null) {
      sender.sendMessage(PREFIX.append(Component.text("Game not found: " + gameId, NamedTextColor.RED)));
      return true;
    }

    Location location = player.getLocation();
    game.setSpawnPoint(location);
    player.sendMessage(PREFIX.append(Component.text("Set spawn point for game: ", NamedTextColor.GREEN))
        .append(Component.text(gameId, NamedTextColor.GOLD)));

    return true;
  }

  private void showHelp(CommandSender sender) {
    sender.sendMessage(Component.text("=== Minigame Command Help ===", NamedTextColor.GOLD));
    sender.sendMessage(Component.text("/minigame join <gameId>", NamedTextColor.YELLOW)
        .append(Component.text(" - Join a minigame", NamedTextColor.WHITE)));
    sender.sendMessage(Component.text("/minigame leave", NamedTextColor.YELLOW)
        .append(Component.text(" - Leave your current minigame", NamedTextColor.WHITE)));
    sender.sendMessage(Component.text("/minigame list", NamedTextColor.YELLOW)
        .append(Component.text(" - List all active minigames", NamedTextColor.WHITE)));

    if (sender.hasPermission("iremiaminigame.command.create")) {
      sender.sendMessage(Component.text("/minigame create <type> <id>", NamedTextColor.YELLOW)
          .append(Component.text(" - Create a new minigame", NamedTextColor.WHITE)));
    }

    if (sender.hasPermission("iremiaminigame.command.start")) {
      sender.sendMessage(Component.text("/minigame start <gameId>", NamedTextColor.YELLOW)
          .append(Component.text(" - Start a minigame", NamedTextColor.WHITE)));
    }

    if (sender.hasPermission("iremiaminigame.command.end")) {
      sender.sendMessage(Component.text("/minigame end <gameId>", NamedTextColor.YELLOW)
          .append(Component.text(" - End a minigame", NamedTextColor.WHITE)));
    }

    if (sender.hasPermission("iremiaminigame.command.setmin")) {
      sender.sendMessage(Component.text("/minigame setmin <gameId> <minPlayers>", NamedTextColor.YELLOW)
          .append(Component.text(" - Set minimum players", NamedTextColor.WHITE)));
    }

    if (sender.hasPermission("iremiaminigame.command.setmax")) {
      sender.sendMessage(Component.text("/minigame setmax <gameId> <maxPlayers>", NamedTextColor.YELLOW)
          .append(Component.text(" - Set maximum players", NamedTextColor.WHITE)));
    }

    if (sender.hasPermission("iremiaminigame.command.setspawn")) {
      sender.sendMessage(Component.text("/minigame setspawn <gameId>", NamedTextColor.YELLOW)
          .append(Component.text(" - Set spawn point to current location", NamedTextColor.WHITE)));
    }
  }
}

File: src/main/kotlin/org/alpacaindustries/iremiaminigamecore/command/MinigameCommand.kt
package org.alpacaindustries.iremiaminigamecore.command

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import org.alpacaindustries.iremiaminigamecore.IremiaMinigameCorePlugin
import org.alpacaindustries.iremiaminigamecore.minigame.AddPlayerResult
import org.alpacaindustries.iremiaminigamecore.minigame.MinigameManager
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

/**
 * Modern Kotlin-based minigame command handler with type-safe command routing.
 *
 * Features:
 * - Sealed class command routing for type safety
 * - Extension functions for cleaner code
 * - Result-based error handling
 * - Improved tab completion
 * - Reduced boilerplate code
 */
class MinigameCommand(private val plugin: IremiaMinigameCorePlugin) : CommandExecutor, TabCompleter {

    private val minigameManager: MinigameManager = plugin.minigameManager!!

    companion object {
        private val miniMessage = MiniMessage.miniMessage()
        private val PREFIX = "<gray>[<gold><b>Minigame</b></gold>] </gray>"
        fun prefixed(msg: String): Component = miniMessage.deserialize(PREFIX + msg)
    }

    init {
        plugin.getCommand("minigame")?.let { command ->
            command.setExecutor(this)
            command.tabCompleter = this
        } ?: plugin.logger.warning("Failed to register minigame command - command not found in plugin.yml")
    }

    /**
     * Sealed class representing all possible minigame subcommands.
     * This provides type safety and exhaustive when expressions.
     */
    sealed class MinigameSubCommand {
        data class Join(val gameId: String) : MinigameSubCommand()
        object Leave : MinigameSubCommand()
        data class Create(val type: String, val id: String) : MinigameSubCommand()
        object List : MinigameSubCommand()
        data class Start(val gameId: String) : MinigameSubCommand()
        data class End(val gameId: String) : MinigameSubCommand()
        data class SetMin(val gameId: String, val minPlayers: Int) : MinigameSubCommand()
        data class SetMax(val gameId: String, val maxPlayers: Int) : MinigameSubCommand()
        data class SetSpawn(val gameId: String) : MinigameSubCommand()
        object Help : MinigameSubCommand()
        data class Unknown(val command: String) : MinigameSubCommand()
    }

    /**
     * Result type for command execution.
     */
    sealed class CommandResult {
        object Success : CommandResult()
        data class Error(val message: Component) : CommandResult()
        data class Usage(val usage: String) : CommandResult()
    }

    /**
     * Permission constants for cleaner code.
     */
    object Permissions {
        const val CREATE = "iremiaminigame.command.create"
        const val START = "iremiaminigame.command.start"
        const val END = "iremiaminigame.command.end"
        const val SET_MIN = "iremiaminigame.command.setmin"
        const val SET_MAX = "iremiaminigame.command.setmax"
        const val SET_SPAWN = "iremiaminigame.command.setspawn"
    }

    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        if (args.isEmpty()) {
            showHelp(sender)
            return true
        }

        return try {
            val subCommand = parseCommand(args)
            val result = executeCommand(sender, subCommand)
            handleResult(sender, result)
            true
        } catch (e: Exception) {
            sender.sendMessage(prefixed("<red>An error occurred: ${e.message}"))
            plugin.logger.warning("Error in command execution: ${e.message}")
            true
        }
    }

    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<String>): List<String> {
        return when (args.size) {
            1 -> getAvailableCommands(sender).filter { it.startsWith(args[0], ignoreCase = true) }
            2 -> getSecondArgCompletions(args[0], args[1])
            3 -> getThirdArgCompletions(args[0], args[1], args[2])
            else -> emptyList()
        }
    }

    /**
     * Parse command arguments into a type-safe command object.
     */
    private fun parseCommand(args: Array<String>): MinigameSubCommand {
        return when (val subCommand = args[0].lowercase()) {
            "join" -> if (args.size >= 2) MinigameSubCommand.Join(args[1]) else MinigameSubCommand.Unknown(subCommand)
            "leave" -> MinigameSubCommand.Leave
            "create" -> if (args.size >= 3) MinigameSubCommand.Create(args[1], args[2]) else MinigameSubCommand.Unknown(subCommand)
            "list" -> MinigameSubCommand.List
            "start" -> if (args.size >= 2) MinigameSubCommand.Start(args[1]) else MinigameSubCommand.Unknown(subCommand)
            "end" -> if (args.size >= 2) MinigameSubCommand.End(args[1]) else MinigameSubCommand.Unknown(subCommand)
            "setmin" -> if (args.size >= 3) {
                args[2].toIntOrNull()?.let { MinigameSubCommand.SetMin(args[1], it) }
                    ?: MinigameSubCommand.Unknown(subCommand)
            } else MinigameSubCommand.Unknown(subCommand)
            "setmax" -> if (args.size >= 3) {
                args[2].toIntOrNull()?.let { MinigameSubCommand.SetMax(args[1], it) }
                    ?: MinigameSubCommand.Unknown(subCommand)
            } else MinigameSubCommand.Unknown(subCommand)
            "setspawn" -> if (args.size >= 2) MinigameSubCommand.SetSpawn(args[1]) else MinigameSubCommand.Unknown(subCommand)
            "help" -> MinigameSubCommand.Help
            else -> MinigameSubCommand.Unknown(subCommand)
        }
    }

    /**
     * Execute a parsed command and return the result.
     */
    private fun executeCommand(sender: CommandSender, command: MinigameSubCommand): CommandResult {
        return when (command) {
            is MinigameSubCommand.Join -> handleJoin(sender, command.gameId)
            is MinigameSubCommand.Leave -> handleLeave(sender)
            is MinigameSubCommand.Create -> handleCreate(sender, command.type, command.id)
            is MinigameSubCommand.List -> handleList(sender)
            is MinigameSubCommand.Start -> handleStart(sender, command.gameId)
            is MinigameSubCommand.End -> handleEnd(sender, command.gameId)
            is MinigameSubCommand.SetMin -> handleSetMin(sender, command.gameId, command.minPlayers)
            is MinigameSubCommand.SetMax -> handleSetMax(sender, command.gameId, command.maxPlayers)
            is MinigameSubCommand.SetSpawn -> handleSetSpawn(sender, command.gameId)
            is MinigameSubCommand.Help -> { showHelp(sender); CommandResult.Success }
            is MinigameSubCommand.Unknown -> handleUnknown(command.command)
        }
    }

    /**
     * Handle command execution results.
     */
    private fun handleResult(sender: CommandSender, result: CommandResult) {
        when (result) {
            is CommandResult.Success -> { /* Already handled in command methods */ }
            is CommandResult.Error -> sender.sendMessage(result.message)
            is CommandResult.Usage -> sender.sendMessage(prefixed(result.usage))
        }
    }

    // Command Handlers

    private fun handleJoin(sender: CommandSender, gameId: String): CommandResult {
        val player = sender.requirePlayer() ?: return CommandResult.Error(
            prefixed("<red>Only players can join games.")
        )

        val game = minigameManager.getActiveGames()[gameId]
        if (game == null) {
            return CommandResult.Error(prefixed("<red>Game not found: <yellow>$gameId"))
        }

        if (game.state == org.alpacaindustries.iremiaminigamecore.minigame.MinigameState.RUNNING && !game.isAllowJoinDuringGame) {
            return CommandResult.Error(prefixed("<red>You cannot join this game while it is running."))
        }

        val addResult = minigameManager.addPlayerToGame(player, gameId)
        when (addResult) {
            is AddPlayerResult.Success -> {
                player.sendMessage(prefixed("<green>You joined the game: <gold>$gameId"))
                return CommandResult.Success
            }

            is AddPlayerResult.AlreadyInGame -> {
                return CommandResult.Error(prefixed("<red>You are already in a game! Leave it first."))
            }

            is AddPlayerResult.GameNotFound -> {
                return CommandResult.Error(prefixed("<red>Game not found: <yellow>$gameId"))
            }

            is AddPlayerResult.Failed -> {
                return CommandResult.Error(prefixed("<red>Failed to join the game."))
            }
        }
    }

    private fun handleLeave(sender: CommandSender): CommandResult {
        val player = sender.requirePlayer() ?: return CommandResult.Error(
            prefixed("<red>Only players can leave games.")
        )

        return if (minigameManager.removePlayerFromGame(player)) {
            player.sendMessage(prefixed("<green>You left the game."))
            CommandResult.Success
        } else {
            CommandResult.Error(prefixed("<red>You are not in a game."))
        }
    }

    private fun handleCreate(sender: CommandSender, gameType: String, gameId: String): CommandResult {
        if (!sender.requirePermission(Permissions.CREATE)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to create minigames."))
        }

        val minigame = minigameManager.createMinigame(gameType, gameId)
        return if (minigame != null) {
            sender.sendMessage(prefixed("<green>Created game: <gold>${minigame.displayName} ($gameId)"))
            CommandResult.Success
        } else {
            CommandResult.Error(prefixed("<red>Failed to create game."))
        }
    }

    private fun handleList(sender: CommandSender): CommandResult {
        val games = minigameManager.getActiveGames()

        if (games.isEmpty()) {
            sender.sendMessage(prefixed("<yellow>There are no active games."))
            return CommandResult.Success
        }

        sender.sendMessage(prefixed("<yellow>Active Games:"))
        games.forEach { (gameId, game) ->
            sender.sendMessage(
                miniMessage.deserialize("<gray>- <gold>$gameId <yellow>(${game.displayName}): <white>${game.playerCount} players, <aqua>Status: ${game.state}")
            )
        }

        return CommandResult.Success
    }

    private fun handleStart(sender: CommandSender, gameId: String): CommandResult {
        if (!sender.requirePermission(Permissions.START)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to start minigames."))
        }

        val game = minigameManager.getActiveGames()[gameId] ?: return CommandResult.Error(
            prefixed("<red>Game not found: <yellow>$gameId")
        )

        game.startCountdown()
        sender.sendMessage(prefixed("<green>Starting countdown for game: <gold>$gameId"))

        return CommandResult.Success
    }

    private fun handleEnd(sender: CommandSender, gameId: String): CommandResult {
        if (!sender.requirePermission(Permissions.END)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to end minigames."))
        }

        val game = minigameManager.getActiveGames()[gameId] ?: return CommandResult.Error(
            prefixed("<red>Game not found: <yellow>$gameId")
        )

        game.end()
        sender.sendMessage(prefixed("<green>Ended game: <gold>$gameId"))

        return CommandResult.Success
    }

    private fun handleSetMin(sender: CommandSender, gameId: String, minPlayers: Int): CommandResult {
        if (!sender.requirePermission(Permissions.SET_MIN)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to modify minigames."))
        }

        val game = minigameManager.getActiveGames()[gameId] ?: return CommandResult.Error(
            prefixed("<red>Game not found: <yellow>$gameId")
        )

        return try {
            game.minPlayers = minPlayers
            sender.sendMessage(prefixed("<green>Set minimum players to $minPlayers for game: <gold>$gameId"))
            CommandResult.Success
        } catch (e: IllegalArgumentException) {
            CommandResult.Error(prefixed("<red>Invalid value: ${e.message}"))
        }
    }

    private fun handleSetMax(sender: CommandSender, gameId: String, maxPlayers: Int): CommandResult {
        if (!sender.requirePermission(Permissions.SET_MAX)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to modify minigames."))
        }

        val game = minigameManager.getActiveGames()[gameId] ?: return CommandResult.Error(
            prefixed("<red>Game not found: <yellow>$gameId")
        )

        return try {
            game.maxPlayers = maxPlayers
            sender.sendMessage(prefixed("<green>Set maximum players to $maxPlayers for game: <gold>$gameId"))
            CommandResult.Success
        } catch (e: IllegalArgumentException) {
            CommandResult.Error(prefixed("<red>Invalid value: ${e.message}"))
        }
    }

    private fun handleSetSpawn(sender: CommandSender, gameId: String): CommandResult {
        if (!sender.requirePermission(Permissions.SET_SPAWN)) {
            return CommandResult.Error(prefixed("<red>You don't have permission to modify minigames."))
        }

        val player = sender.requirePlayer() ?: return CommandResult.Error(
            prefixed("<red>Only players can set spawn points.")
        )

        val game = minigameManager.getActiveGames()[gameId] ?: return CommandResult.Error(
            prefixed("<red>Game not found: <yellow>$gameId")
        )

        game.spawnPoint = player.location
        sender.sendMessage(prefixed("<green>Set spawn point for game: <gold>$gameId"))

        return CommandResult.Success
    }

    private fun handleUnknown(command: String): CommandResult {
        return CommandResult.Error(prefixed("<red>Unknown command: $command"))
    }

    // Tab Completion Helpers

    private fun getAvailableCommands(sender: CommandSender): List<String> {
        val commands = mutableListOf("join", "leave", "list", "help")

        if (sender.hasPermission(Permissions.CREATE)) commands.add("create")
        if (sender.hasPermission(Permissions.START)) commands.add("start")
        if (sender.hasPermission(Permissions.END)) commands.add("end")
        if (sender.hasPermission(Permissions.SET_MIN)) commands.add("setmin")
        if (sender.hasPermission(Permissions.SET_MAX)) commands.add("setmax")
        if (sender.hasPermission(Permissions.SET_SPAWN)) commands.add("setspawn")

        return commands
    }

    private fun getSecondArgCompletions(firstArg: String, prefix: String): List<String> {
        return when (firstArg.lowercase()) {
            "join", "start", "end", "setmin", "setmax", "setspawn" -> {
                minigameManager.getActiveGames().keys.filter { it.startsWith(prefix, ignoreCase = true) }
            }
            "create" -> {
                minigameManager.getMinigameTypes().filter { it.startsWith(prefix, ignoreCase = true) }
            }
            else -> emptyList()
        }
    }

    private fun getThirdArgCompletions(firstArg: String, secondArg: String, prefix: String): List<String> {
        return when (firstArg.lowercase()) {
            "create" -> {
                listOf("game1", "game2", "custom", "${secondArg}_1")
                    .filter { it.startsWith(prefix, ignoreCase = true) }
            }
            "setmin", "setmax" -> {
                listOf("1", "2", "4", "8", "16", "32")
                    .filter { it.startsWith(prefix, ignoreCase = true) }
            }
            else -> emptyList()
        }
    }

    // Help System

    private fun showHelp(sender: CommandSender) {
        sender.sendMessage(prefixed("<yellow>Available Commands:"))

        val helpEntries = listOf(
            "/minigame join <gameId>" to "Join a minigame",
            "/minigame leave" to "Leave your current minigame",
            "/minigame list" to "List all active minigames"
        ).toMutableList()

        if (sender.hasPermission(Permissions.CREATE)) {
            helpEntries.add("/minigame create <type> <id>" to "Create a new minigame")
        }
        if (sender.hasPermission(Permissions.START)) {
            helpEntries.add("/minigame start <gameId>" to "Start a minigame countdown")
        }
        if (sender.hasPermission(Permissions.END)) {
            helpEntries.add("/minigame end <gameId>" to "End a minigame")
        }
        if (sender.hasPermission(Permissions.SET_MIN)) {
            helpEntries.add("/minigame setmin <gameId> <count>" to "Set minimum players")
        }
        if (sender.hasPermission(Permissions.SET_MAX)) {
            helpEntries.add("/minigame setmax <gameId> <count>" to "Set maximum players")
        }
        if (sender.hasPermission(Permissions.SET_SPAWN)) {
            helpEntries.add("/minigame setspawn <gameId>" to "Set spawn point to your location")
        }

        helpEntries.forEach { (command, description) ->
            sender.sendMessage(miniMessage.deserialize("<gold>  $command <gray>- $description"))
        }
    }
}

// Extension Functions for Cleaner Code

private fun CommandSender.requirePlayer(): Player? = this as? Player

private fun CommandSender.requirePermission(permission: String): Boolean {
    return hasPermission(permission).also {
        if (!it) sendMessage(MinigameCommand.prefixed("<red>You don't have permission."))
    }
}
